tests/datasets/test_streaming.py

```diff
@@ -0,0 +1,163 @@
+
+    first_sample_values = samples[0].values()
+    first_sample_byte_sizes = np.array([len(v) for v in first_sample_values], dtype=np.int64)
+    first_sample_bytes = len(first_sample_byte_sizes.tobytes() + b''.join(first_sample_values))
+
+    expected_samples_per_shard = shard_size_limit // first_sample_bytes
+    expected_num_shards = math.ceil(num_samples / expected_samples_per_shard)
+    expected_num_files = expected_num_shards + 1  # the index file
+
+    dirname = write_synthetic_streaming_dataset(samples=samples, shard_size_limit=shard_size_limit)
+    files = os.listdir(dirname)
+
+    assert len(files) == expected_num_files, f"Files written ({len(files)}) != expected ({expected_num_files})."
+
+
+@pytest.mark.parametrize("share_remote_local", [False, True])
+@pytest.mark.parametrize("shuffle", [False, True])
+def test_reader(share_remote_local: bool, shuffle: bool):
+    num_samples = 117
+    shard_size_limit = 1 << 8
+    samples, decoders = get_fake_samples_decoders(num_samples)
+    remote = write_synthetic_streaming_dataset(samples=samples, shard_size_limit=shard_size_limit)
+    if share_remote_local:
+        local = remote
+    else:
+        local = tempfile.TemporaryDirectory().name
+
+    # Build StreamingDataset
+    dataset = StreamingDataset(remote=remote, local=local, shuffle=shuffle, decoders=decoders)
+
+    # Test length
+    assert len(dataset) == num_samples, f"Got dataset length={len(dataset)} samples, expected {num_samples}"
+
+    # Test basic sample order
+    shuffle_matches = 0
+    for ix, sample in enumerate(dataset):
+        uid = sample["uid"]
+        data = sample["data"]
+        expected_uid = f"{ix:06}"
+        expected_data = 3 * ix
+        if shuffle:
+            shuffle_matches += (expected_uid == uid)
+        else:
+            assert uid == expected_uid == uid, f"sample ix={ix} has uid={uid}, expected {expected_uid}"
+            assert data == expected_data, f"sample ix={ix} has data={data}, expected {expected_data}"
+
+    # If shuffling, there should be few matches
+    # The probability of k matches in a random permutation is ~1/(e*(k!))
+    if shuffle:
+        assert shuffle_matches < 10
+
+
+@pytest.mark.timeout(10)
+@pytest.mark.parametrize("created_ago", [0.5, 3])
+@pytest.mark.parametrize("timeout", [1])
+def test_reader_after_crash(created_ago, timeout):
+    num_samples = 117
+    shard_size_limit = 1 << 8
+    samples, decoders = get_fake_samples_decoders(num_samples)
+    remote = write_synthetic_streaming_dataset(samples=samples, shard_size_limit=shard_size_limit)
+    local = tempfile.TemporaryDirectory().name
+
+    os.makedirs(local, exist_ok=True)
+    shutil.copy(os.path.join(remote, "index.mds"), os.path.join(local, "index.mds.tmp"))
+    shutil.copy(os.path.join(remote, "000003.mds"), os.path.join(local, "000003.mds.tmp"))
+    time.sleep(created_ago)
+    dataset = StreamingDataset(remote=remote, local=local, shuffle=False, decoders=decoders, timeout=timeout)
```

Comment:
Is it possible to test multiple crashes? (See comment in the function about calling the wait recursively...)

Reply:
I elaborated more on this in the download function, but I interpret timeout as a "per attempt" timeout, which I think should be separate from the number of retries.

Reply:
Tracking this as part of a broader timeout refactor in JIRA.